### PR TITLE
[IMP] web_editor, website: improve text highlight system

### PR DESCRIPTION
**Pull request by xO-Tx**
The goal of this PR is to improve the current text highlight system by adding
the following updates:

- Move the highlight option's related CSS ([data-set-text-highlight]) to website.

- Update the `selectStyle` options to handle the default "currentColor" value.

- Force to hide the highlight options overlay opacity on scroll.

- Remove the dead code of `getDOMRectWidth()` (used in older versions to compare
text content width on page load).

- The `o_text_highlight_item_dirty` class is no longer used to adapt non-detected
highlight items (the highlight mutation observer was tweaked to detect and adapt
the new highlight items added to the DOM after editors commands. E.g. after
the `oEnter()` command).

- The editor's code (`_handleTextOptions()`) will only set text highlight classes and
activate the snippet. Highlight DOM adaptations (multi-line detection, setting
SVGs...) are done by the public widget.

- A `z-index:` CSS (used to keep the highlight SVG visible and behind the text at the
same time) is replaced by a more appropriate `isolation: isolate;`.

- Only use one global `resizeObserver` to adapt the text highlights on resize.

- The text highlight effects list was updated.

task-3285817

### CODE REVIEW

**Comment by xO-Tx** [comment]
PING @qsm-odoo for review. Just let me know if something needs to be adjusted here :+1:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -23,7 +23,6 @@ import {
 import { renderToElement, renderToFragment } from "@web/core/utils/render";
 import { browser } from "@web/core/browser/browser";
 import {
-    applyTextHighlight,
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:26]: Bug: you are removing this while it is used by this commit.

File: addons/website/static/src/js/text_processing.js [hunk 2]
```diff
@@ -181,8 +181,6 @@ const _textHighlightFactory = {
         return drawPath(targetEl, {mode: "fill", template, SVGWidth: 200, SVGHeight: 43});
     },
 };
-// Returns the width of the DOMRect object.
-export const getDOMRectWidth = el => el.getBoundingClientRect().width;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/text_processing.js:185]: Let's not do that one in 17.0, move this out in another PR for master.

File: addons/website/static/src/js/text_processing.js [hunk 3]
```diff
@@ -5,9 +5,6 @@ import * as OdooEditorLib from "@web_editor/js/editor/odoo-editor/src/utils/util
 // SVG generator: contains all information needed to draw highlight SVGs
 // according to text dimensions, highlight style,...
 const _textHighlightFactory = {
-    underline: targetEl => {
-        return drawPath(targetEl, {mode: "line"});
-    },
```
Comment (qsm-odoo) [inline addons/website/static/src/js/text_processing.js:10]: With this commit as it is, if a current 17.0 user did use that effect, it will now crash on every page load... we of course cannot merge that in 17.0 at this point :grimacing: 

1. Let's move this commit for master as well, it's too late for 17.0
2. We now have to consider migration / compatibility

File: addons/website/static/src/js/editor/snippets.editor.js [hunk 4]
```diff
@@ -473,29 +468,6 @@ const wSnippetMenu = weSnippetEditor.SnippetsMenu.extend({
...
         targetEl.classList.remove('active');
     },
-    /**
-     * Used to adjust the highlight effect when the text content is edited.
-     *
-     * @private
-     * @param {HTMLElement} target
-     * @param {String} highlightID
-     */
-    _adaptHighlightOnEdit(target, highlightID) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.editor.js:483]: Let's keep the various refactoring this commit does in stable so we do not have different codebase that close to 17.0... but let's be a bit more careful: keep existing function, just stop using them and mark them to be removed in master.

File: addons/website/static/src/js/editor/snippets.editor.js [hunk 5]
```diff
@@ -621,11 +624,17 @@ const wSnippetMenu = weSnippetEditor.SnippetsMenu.extend({
...
+        // the `_closeWidgets()` method) right after opening it.
+        ev.stopPropagation();
+        this._closeWidgets();
         const target = ev.currentTarget;
         this._handleTextOptions(target, [
             this._getOptionTextClass(target),
             "o_text_highlight_freehand_1",
-        ]);
+        ],
+        () => this.options.wysiwyg.odooEditor.dispatchEvent(new Event("show_text_highlights")));
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.editor.js:637]: - Strange alignment IMO
- Maybe you could use the system around `_onOptionUpdate` instead of disaptching event through the odooEditor which is really not related to snippet options?

File: addons/website/static/src/js/content/snippets.animation.js [hunk 6]
```diff
@@ -1837,131 +1837,146 @@ registry.TextHighlight = publicWidget.Widget.extend({
...
+        this._super(...arguments);
+        // We need to adapt the text highlights on resize (E.g. custom fonts
+        // loading, layout option changes, window resized...), mainly to take in
+        // consideration the rendered line breaks in text nodes... But after
+        // every adjustment, the `ResizeObserver` will unfortunately immediately
+        // notify a size change once new highlight items are observed leading to
+        // an infinite loop. To avoid that, we use a lock map (`observerLock`)
+        // to block the callback on this first notification for observed items.
+        this.observerLock = new Map();
+        this.resizeObserver = new window.ResizeObserver(entries => {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js:1853]: Do you actually need to declare this in the `init` instead of the start as before? It's ok but in general, using start for public widgets is more conventional.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I reviewed this very quickly, I'll have a deeper review once this is properly split between 17.0 and master. We can keep most of the refactoring for 17.0 for this one, as it will be quite annoying to maintain 2 different codebase that close after the 17.0 release but make sure this is stable: existing users should not lose their highlight effect, it should not crash for them, functions should not be removed, etc etc :+1:

File: addons/website/static/src/js/text_processing.js (hunk 3 above)
Comment (xO-Tx) [reply addons/website/static/src/js/text_processing.js]: Will be done in https://github.com/odoo/odoo/pull/149820 :+1:

File: addons/website/static/src/js/editor/snippets.editor.js (hunk 4 above)
Comment (xO-Tx) [reply addons/website/static/src/js/editor/snippets.editor.js:505]: OK :+1: I did this for other functions, but for this one, I think it can be removed since [there is already a code to set it in start()](https://github.com/odoo/odoo/pull/142968/files#diff-d2188391a9d83cc97f3220e08d259e82796d94191dcf5e5fdb9b77e57074e6a5R88), no?

File: addons/website/static/src/js/editor/snippets.editor.js (hunk 5 above)
Comment (xO-Tx) [reply addons/website/static/src/js/editor/snippets.editor.js]: > Maybe you could use the system around _onOptionUpdate instead of disaptching event through the odooEditor which is really not related to snippet options?

Yes, we can trigger an `"option_update"` from the editor linked to the activated text snippet.
Actually, I had to do it that way because it's only on `_handleTextOptions()` where I'm able to check if the highlight button was clicked for the first time in the toolbar, and also the system using the `odooEditor` event to link it with the options was already implemented in the same way to [activate link tools](https://github.com/odoo/odoo/blob/17.0/addons/web_editor/static/src/js/editor/snippets.options.js#L5901).

I updated the code now to use the `"option_update"` & `notify()` systems, thanks :+1:

File: addons/website/static/src/js/editor/snippets.editor.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.editor.js:505]: Let's just not take the risk. Actually maybe just keep this line: `this._adaptHighlightOnEdit = throttleForAnimation(this._adaptHighlightOnEdit.bind(this));` and here in the definition of `_adaptHighlightOnEdit` just call `switchTextHighlight`, then in master you'll simplify during forward-port.

FYI, even if defined like that in the `start`, any custo can still break it... simply by calling it before start by mistake. It's wrong if someone did that... but their code would not crash and would start crashing after merging this PR :grimacing:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@xO-Tx This seems mostly nice. I push-forced with some changes: https://github.com/odoo/odoo/compare/e7a76e3c198316533a10671b5a65a0295612664c..9dff32ed7670f39f16bf63b602099450001eb9f9

Some nitpicking but also some changes to make it more stable. Some explanations:
- `_adaptHighlightOnEdit` second argument was now useless
- Chose to not remove `applyTextOption` argument in stable: let's be cautious
- Used `__handleTextOptionsPostActivate` to not modify the `_handleTextOptions` arguments in stable
- No change in arguments for `selectElementInWeSelectWidget`. I think we should consider that developers can use our website utils to write tests. We should not have all our community developers tests red all of a sudden because we changed the way the utils are defined in stable :wink: 

I reviewed all commits but the last one. Just before starting to review the last one, I found a bug, due to this branch. It was there before my push-force. You can see it here: https://57766741-17-0-all.runbot201.odoo.com/ (this is the runbot before my push-force)
=> There is some content with highlights on the homepage
=> Select the whole snippet, try to delete it => many tracebacks.

With the exact same content in 17.0, it works.

Could you fetch (to get my changes) and investigate and fix that? :pray: Maybe re-try everything to be sure too :+1: After that I'll review the last commit and merge :pray:

**Comment by xO-Tx** [comment]
> I reviewed all commits but the last one. Just before starting to review the last one, I found a bug, due to this branch. It was there before my push-force. You can see it here: https://57766741-17-0-all.runbot201.odoo.com/ (this is the runbot before my push-force)

@qsm-odoo I took a look at this bug, and it seems caused by a remaining `start() > _autoAdaptHighlights()` on the `TextHighlight` snippet option (used to reapply the effects by refreshing the highlight public widget), so what happens exactly on snippet removal:

- We call `removeSnippet()`, that triggers the `"call_for_each_child_snippet"` to call `onRemove` of all internal options (including highlights). Leading to execute the `_createSnippetEditor()`.
- With the public widgets restart (`widgets_start_request`) triggered at every `start()` of highlight snippets, this will lead to inconsistencies on `destroy()` of public widgets that target the same DOM. 

Actually, the highlight effect DOM structure should be applied on every public widget restart **[A]** and when a new highlight effect is added **[B]** and having the refresh code on every `start()` is not suitable to just handle the **[B]** case.

Code updated now: I moved the `_autoAdaptHighlights()` to the `notify` / `__handleTextOptionsPostActivate` code (to only add highlights when applied for the first time) the same way we do to display the effects grid. Just let me know if something needs to be adjusted here :+1:

File: addons/website/static/src/js/content/snippets.animation.js [hunk 7]
```diff
@@ -1879,44 +1879,65 @@ registry.TextHighlight = publicWidget.Widget.extend({
...
+                            return this.observerLock.set(target, false);
+                        }
+                        topTextEls.add(topTextEl);
+                        if (!target.classList.contains("o_text_highlight")) {
+                            // Unobserve the target (it will be replaced by
+                            // a new one after the update).
+                            this.resizeObserver.unobserve(target);
+                        }
+                    }
+                });
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js:1910]: I don't understand this code.
- You have a `target` which is the element who was being observed and thus which got resized
- You search for the closest text_hightlight element ancestors find then inside the target itself
- If there is one, you unlock the target... not the text_highlight element you found? While the `_lockHighlightObserver` seems to lock only element which are text_highlight themselves
- Then if the target is not the text_highlight element itself, you stop observing it... which case is that? A element which was a text_highlight, got obserbed but somehow now contains or is in another text hightlight but is not a text highlight itself? :thinking:

File: addons/website/static/src/js/content/snippets.animation.js [hunk 8]
```diff
@@ -1879,44 +1879,65 @@ registry.TextHighlight = publicWidget.Widget.extend({
...
-            for (const textHighlightEl of this.el.querySelectorAll(".o_text_highlight")) {
-                removeTextHighlight(textHighlightEl);
-            }
+        // We only save the highlight information on the main text wrapper,
+        // the full structure will be restored on page load.
+        for (const textHighlightEl of this.el.querySelectorAll(".o_text_highlight")) {
+            removeTextHighlight(textHighlightEl);
         }
+        this.resizeObserver.disconnect();
+        this.observerLock.clear();
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js:1939]: This is not useful, the widget is being destroyed anyway

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some last comments? :muscle:

File: addons/website/static/src/js/content/snippets.animation.js (hunk 7 above)
Comment (xO-Tx) [reply addons/website/static/src/js/content/snippets.animation.js]: @qsm-odoo Here is a summary of the `resizeObserver`'s logic (after adding some updates & fixes):

**[1]- First step (on public widget `start()`):**

A- The highlights are applied > `applyTextHighlight()`;

B- For every `.o_text_highlight`, its parent **[x]** and its `.o_text_highlight_item` elements are locked > `_lockHighlightObserver()`

C- For every `.o_text_highlight`, its parent **[x]** and its `.o_text_highlight_item` elements are observed > `_observeHighlightResize()`

D- Right after the `resizeObserver.observe()`, [the resize handler will be called automatically even if there is no "resize"](https://drafts.csswg.org/resize-observer/#intro).

E- This first handler call should just unlock "observed targets":
```javascript
if (this.observerLock.get(target)) {
    return this.observerLock.set(target, false);
}
```
**[x]:** Fixed: The `resizeObserver` will simply ignore an element if it has an inline display. We should actually target the closest non-inline parent (see: `_closestToObserve()`).

**[2]- On every "RESIZE":**

A- Resize handler should get the `.o_text_highlight` items (inside the target if it's a parent or the closest one if the target is a highlight item...) and add them to the adaptation `Set()`.

B- We unobserve all highlight lines before the update since they will be automatically removed from the DOM (because of highlight lines computing...).

C- After `switchTextHighlight()`, we will automatically go to [1.B].

To respond to your questions:

> You have a `target` which is the element who was being observed and thus which got resized
> You search for the closest text_hightlight element ancestors find then inside the target itself
- If the target is a parent, we need to get the targeted ".o_text_highlight" elements **[xx]**, otherwise, it's the `closest()` one (see: [1.C]).

**[xx]**: Fixed: If the target is a parent, we should get all highlight elements, not just the first one.

> If there is one, you unlock the target... not the text_highlight element you found? While the _lockHighlightObserver seems to lock only element which are text_highlight themselves
- `_lockHighlightObserver()` locks all observed targets, (see: [1.D-E]) 

> Then if the target is not the text_highlight element itself, you stop observing it... which case is that? A element which was a text_highlight,
got obserbed but somehow now contains or is in another text hightlight but is not a text highlight itself? 🤔
- Good catch! Actually, the condition was from an old code version where only the content inside `.o_text_highlight` was observed and it was used to target the `.o_text_highlight_item` lines (see: [2.B]). Fixed now :+1:

**Remark:** I added a separate commit to allow the translation of text options (text animation & highlights) in stable as a fix for **opw-3686777**. Just let me know if something needs to be adjusted here :+1:

Comment (qsm-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: @xO-Tx Thanks :+1: I quickly flew over the code, I'll not check all details again, seems good :+1: 

> Remark: I added a separate commit to allow the translation of text options (text animation & highlights) in stable as a fix for opw-3686777. Just let me know if something needs to be adjusted here 👍

Could you move that to its own PR? I'd like to comment some things on that one. Also, should not it be for 16.0 in the end? But maybe we can indeed just hide the text animation button in 16.0 and take the risk in 17.0 only.

Comment (xO-Tx) [reply addons/website/static/src/js/content/snippets.animation.js]: > Could you move that to its own PR? I'd like to comment some things on that one. Also, should not it be for 16.0 in the end? But maybe we can indeed just hide the text animation button in 16.0 and take the risk in 17.0 only.

The code will be added in the `17.0` forward port of: https://github.com/odoo/odoo/pull/157087. It targets `15.0` now (where the "Text Animations" were available) and, as suggested, the option button will be disabled for `15.0~16.0`.
I also fixed another issue that makes animated text invisible in "translate" mode in the same PR. If something needs to be changed, just let me know :+1:

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+ :partying_face:
